src/action-providers.ts

```diff
@@ -21,9 +21,73 @@ class RefactoringActionProvider implements vscode.CodeActionProvider {
     const ast = t.parse(document.getText());
     const selection = createSelectionFromVSCode(range);
 
-    return this.refactorings
-      .filter(refactoring => this.canPerform(refactoring, ast, selection))
-      .map(refactoring => this.buildCodeActionFor(refactoring));
+    const applicableRefactorings: RefactoringWithActionProvider[] = [];
+    t.traverseAST(ast, {
+      enter: (path: t.NodePath<any>) => {
+        this.refactorings.forEach(
+          this.checkCanPerform(
+            path,
+            selection,
+            (
+              matchedPath: t.NodePath<any>,
+              refactoring: RefactoringWithActionProvider
+            ) => {
+              if (refactoring.actionProvider.canPerformRefactoringMutator) {
+                applicableRefactorings.push(
+                  refactoring.actionProvider.canPerformRefactoringMutator(
+                    matchedPath,
+                    refactoring
+                  )
+                );
+              } else {
+                applicableRefactorings.push(refactoring);
+              }
+            }
+          )
+        );
+        path.type;
+      }
+    });
+
+    return [
+      ...applicableRefactorings.map(refactoring =>
+        this.buildCodeActionFor(refactoring)
+      ),
+      ...this.refactorings
+        .filter(refactoring => this.canPerform(refactoring, ast, selection))
+        .map(refactoring => this.buildCodeActionFor(refactoring))
+    ];
+  }
+
+  private checkCanPerform(
+    path: any,
+    selection: any,
+    onCanPerform: (
+      matchedPath: t.NodePath<any>,
+      refactoring: RefactoringWithActionProvider
+    ) => void
+  ) {
+    return (refactoring: RefactoringWithActionProvider) => {
```

Comment:
Maybe we could simplify this and not return a function from this method. I think this can be done by pulling up the `(refactoring: RefactoringWithActionProvider) => {}` back in the `forEach()` caller. What do you think?

Reply:
Yes, that works. It will be part of the next proposal.